core/domain/user_services.py

```diff
@@ -2287,3 +2291,257 @@ def get_dashboard_stats(user_id):
+            _get_most_distant_reached_checkpoint_in_current_exploration(
+                checkpoints_in_current_exp,
+                checkpoints_in_older_exp,
+                exp_user_model.furthest_reached_checkpoint_state_name
+            )
+        )
+
+        # If the furthest reached checkpoint doesn't exist in current
+        # exploration.
+        if furthest_reached_checkpoint_in_current_exp is None:
+            exp_user_model.furthest_reached_checkpoint_exp_version = (
+                exp_version)
+            exp_user_model.furthest_reached_checkpoint_state_name = state_name
+        else:
+            # Index of the furthest reached checkpoint.
+            frc_index = checkpoints_in_current_exp.index(
+                furthest_reached_checkpoint_in_current_exp)
+            # If furthest reached checkpoint is behind most recently
+            # reached checkpoint.
+            if frc_index <= checkpoints_in_current_exp.index(state_name):
+                exp_user_model.furthest_reached_checkpoint_exp_version = (
+                    exp_version)
+                exp_user_model.furthest_reached_checkpoint_state_name = (
+                    state_name)
+
+    exp_user_model.most_recently_reached_checkpoint_exp_version = exp_version
+    exp_user_model.most_recently_reached_checkpoint_state_name = state_name
+    exp_user_model.update_timestamps()
+    exp_user_model.put()
+
+
+def set_user_has_viewed_lesson_info_modal_once(user_id):
+    """Updates the user's settings once he has viewed the lesson info modal.
+
+    Args:
+        user_id: str. The Id of the user.
+    """
+    user_settings = get_user_settings(user_id)
+    user_settings.mark_lesson_info_modal_viewed()
+    _save_user_settings(user_settings)
+
+
+def update_learner_checkpoint_progress_on_restart(user_id, exploration_id):
+    """Sets the most recently reached checkpoint on restart exploration event.
+
+    Args:
+        user_id: str. The Id of the user.
+        exploration_id: str. The Id of the exploration.
+    """
+    exp_user_model = user_models.ExplorationUserDataModel.get(
+        user_id, exploration_id)
+    exp_user_model.most_recently_reached_checkpoint_exp_version = None
+    exp_user_model.most_recently_reached_checkpoint_state_name = None
+    exp_user_model.update_timestamps()
+    exp_user_model.put()
+
+
+def sync_learner_checkpoint_progress_with_current_exp_version(
+    user_id,
+    exploration_id
+):
+    """Synchronizes the most recently reached checkpoint and the furthest
+    reached checkpoint with the latest exploration.
+
+    Args:
+        user_id: str. The Id of the user.
+        exploration_id: str. The Id of the exploration.
+
+    Returns:
+        ExplorationUserData. The domain object corresponding to the given user
+        and exploration.
+    """
+    exp_user_model = user_models.ExplorationUserDataModel.get(
+        user_id, exploration_id)
+
+    if exp_user_model is not None:
```

Comment:
```suggestion
    if exp_user_model is None:
        return None
```

and deintend the code below

try to do this in other parts of the code too

Reply:
Done!